controllers/remotesecret_controller.go

```diff
@@ -157,12 +157,22 @@ func (r *RemoteSecretReconciler) Reconcile(ctx context.Context, req reconcile.Re
 	}
 
 	// the reconciliation happens in stages, results of which are described in the status conditions.
-
-	dataResult, err := handleStage(ctx, r.Client, remoteSecret, r.obtainData(ctx, remoteSecret))
+	var secretData stageResult[*remotesecretstorage.SecretData]
+	secretData = r.obtainData(ctx, remoteSecret)
+	dataResult, err := handleStage(ctx, r.Client, remoteSecret, secretData)
 	if err != nil || dataResult.Cancellation.Cancel {
 		return dataResult.Cancellation.Result, err
 	}
 
+	remoteSecret.Status.Secret.Keys = make([]string, 0)
+	for k := range *secretData.ReturnValue {
+		remoteSecret.Status.Secret.Keys = append(remoteSecret.Status.Secret.Keys, k)
+	}
+
+	if err := r.Client.Status().Update(ctx, remoteSecret); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update the status with a list of keys in storage: %w", err)
+	}
```

Comment:
Wouldn't it make more sense if the update of the status with the keys was also part of the obtainData stage? You could avoid the explicit update of the status there because the status is persisted right at the end of the stage.